guildwars2/key.py

```diff
@@ -84,43 +111,159 @@ class KeyMixin:
         """
-        user = ctx.author
+        if ctx.author.id in KeyMixin.waitingfor:
+            await ctx.send("I'm already waiting for a response from you for another key command.")
+            return
+        doc = await self.bot.database.get_user(ctx.author, self)
         try:
-            await self.fetch_key(user)
-        except:
-            return await ctx.send(
-                "{.mention}, no API key associated with your account. "
-                "Add your key using `$key add` command.".format(user))
-        await self.bot.database.set_user(user, {"key": {}}, self)
-        await ctx.send("{.mention}, sucessfuly removed your key. "
-                       "You may input a new one.".format(user))
+            keys = doc["keys"]
+        except KeyError:
+            #no keys
+            keys = []
+            try:
+                key = doc["key"]
+                key = {}
+            except KeyError:
+                return await ctx.send("You have no keys added, you can add one with {0}key add.".format(ctx.prefix))
+        else:
+            key = doc["key"]
+        if doc["keys"] == [] and doc["key"] == {}:
+            return await ctx.send("You have no keys added, you can add one with {0}key add.".format(ctx.prefix))
+        if len(keys) > 0:
+            if ctx.guild is not None:
+                await ctx.send("Check your PMs.")        
+            output = "Type the number of the key you wish to delete, or respond with all to remove all keys.\n```"
+            for count, key in enumerate(keys, 1):
+                output += str(count) + ": " + key["name"] + " Account: " + key["account_name"] + " Key: " + key["key"] +"\n"
+            output += "```"
+            message = await ctx.author.send(output)
+            def check(m):
+                return m.author == ctx.author and m.channel == message.channel
+            try:
+                KeyMixin.waitingfor.append(ctx.author.id)
+                answer = await self.bot.wait_for("message", timeout=120, check=check)
+            except asyncio.TimeoutError:
+                await message.edit(content="No response in time.")
+                return
+            finally:
+                KeyMixin.waitingfor.remove(ctx.author.id)
+
+            try:
+                num = int(answer.content) - 1
+                if keys[num]["key"] == doc["key"]["key"]:
+                    key = {}
+                    await ctx.author.send("This was your active key, you won't be able to use "
+                        "commands that require a key unless you set a new key with +key active.")
+                del keys[num]
+            except:
+                if 'all' in answer.content.lower():
+                    keys = []
+                    key = {}
+                else:
+                    await message.edit(content="That's not a number in the list.")
+                    return
+        await self.bot.database.set_user(ctx.author, {"key": key, "keys": keys}, self)
```

Comment:
Might want to look into push/pull operators in mongodb to simplify it a bit, https://docs.mongodb.com/manual/reference/operator/update/pull/
worldpoptrack uses it